### PR TITLE
Dockerfile: go mod cache

### PR DESCRIPTION
**Pull request by 7sunarni**
cache go modules, so improve building speed
https://docs.docker.com/language/golang/build-images/#create-a-dockerfile-for-the-application

### CODE REVIEW

**Comment by 7sunarni** [comment]
assign @thor-wl

**Comment by 7sunarni** [comment]
Hi~ Can you take minutes review this, it could speed up CI stage
@william-wang

**Comment by hwdef** [comment]
please make the DCO CI happy first

**Comment by 7sunarni** [comment]
> please make the DCO CI happy first

Done

**Comment by 7sunarni** [comment]
It seems both of reviewers not active any more, how about assign new reviewer?

**Comment by 7sunarni** [comment]
Update: [this commit](https://github.com/volcano-sh/volcano/pull/3142/commits/a32b344f9990b106f6317d0449c2bcb9b13c44ba) use alpine mirror to speed up building webhook-manager docker image.

**Comment by 7sunarni** [comment]
@Monokaix Hi~ Sorry to tag you. It looks like that two reviewers assigned not active in this project. Do you know how and who to assign new reviewers?

**Comment by Monokaix** [comment]
> Update: [this commit](https://github.com/volcano-sh/volcano/pull/3142/commits/a32b344f9990b106f6317d0449c2bcb9b13c44ba) use alpine mirror to speed up building webhook-manager docker image. I think most of Volcano's user in mainland. This mirror could be useful in improving building container image.

It's a good intention here. But volcano is used all over the world, we should not specify cdn url in a hard code way, my opinion is that set a parameter like env to let user specify custom cdn url or just disable it: )

**Comment by Monokaix** [comment]
Welcome! A little doubt here. Does this cache optimization you mentioned only take effect when building multiple binaries? Seems that when we execute `make vc-controller-manager`, the `go mod download` will be executed, and after your pr, this step has been moved to the front, so what's the difference here?

**Comment by 7sunarni** [comment]
> Welcome! A little doubt here. Does this cache optimization you mentioned only take effect when building multiple binaries? 
I thinks this will work when you need to build the image in high frequency. It will not download the dependency each time if you don't modify the go.mod and go.sum.

**Comment by 7sunarni** [comment]
>  Does this cache optimization you mentioned only take effect when building multiple binaries? 

I think when your change not including go.mod and go.sum file and just including some go file, in docker building stage, this will not download dependency. Not only in building multiple binaries, and in local develop/test stage building image. 

> Seems that when we execute make vc-controller-manager, the go mod download will be executed, and after your pr, this step has been moved to the front, so what's the difference here?

As I know, docker will use the cache if the file don't change. In most case, `go.mod` and `go.sum` will not change,  so docker will use cache in step3 and step4 . This is also what [Docker document](https://docs.docker.com/language/golang/build-images/#multi-stage-builds) recommend us to do. And if we just copy `.volcano` to the build, this step has file change, docker will not use cache from this.

File: installer/dockerfile/webhook-manager/Dockerfile [hunk 1]
```diff
@@ -14,13 +14,17 @@
...
+RUN go mod download
 ADD . volcano
 RUN cd volcano && make vc-webhook-manager
 
 FROM alpine:latest
 ARG KUBE_VERSION="1.27.0"
 ARG TARGETARCH
-RUN apk add --update ca-certificates && \
+ARG APK_MIRROR
+RUN if [[ -n "$APK_MIRROR" ]]; then sed -i "s@https://dl-cdn.alpinelinux.org@${APK_MIRROR}@g" /etc/apk/repositories ; fi && \
```
Comment (Monokaix) [inline installer/dockerfile/webhook-manager/Dockerfile:26]: I think we should also add --build-args includes this env in docker build command: )

**Comment by 7sunarni** [comment]
> > Update: [this commit](https://github.com/volcano-sh/volcano/pull/3142/commits/a32b344f9990b106f6317d0449c2bcb9b13c44ba) use alpine mirror to speed up building webhook-manager docker image. I think most of Volcano's user in mainland. This mirror could be useful in improving building container image.
> 
> It's a good intention here. But volcano is used all over the world, we should not specify cdn url in a hard code way, my opinion is that set a parameter like env to let user specify custom cdn url or just disable it: )

You are right. I add a argument like `--build-arg APK_MIRROR="https://mirrors.huaweicloud.com"`  to let users do their choice, default will not do anything. 

By the way, it seems you are from Huawei. And the usage in this [official article](https://mirrors.huaweicloud.com/mirrorDetail/5ea14e7b02ea0eda6bd1e8fa) seems have a little mistake. The config command should be `sed -i "s@https://dl-cdn.alpinelinux.org/@https://repo.huaweicloud.com/@g" /etc/apk/repositories` . The comments in this article had point out this mistake, but not correct yet. :-(

File: installer/dockerfile/webhook-manager/Dockerfile (hunk 1 above)
Comment (7sunarni) [reply installer/dockerfile/webhook-manager/Dockerfile:26]: You are right.

**Comment by Monokaix** [comment]
> > > Update: [this commit](https://github.com/volcano-sh/volcano/pull/3142/commits/a32b344f9990b106f6317d0449c2bcb9b13c44ba) use alpine mirror to speed up building webhook-manager docker image. I think most of Volcano's user in mainland. This mirror could be useful in improving building container image.
> > 
> > 
> > It's a good intention here. But volcano is used all over the world, we should not specify cdn url in a hard code way, my opinion is that set a parameter like env to let user specify custom cdn url or just disable it: )
> 
> You are right. I add a argument like `--build-arg APK_MIRROR="https://mirrors.huaweicloud.com"` to let users do their choice, default will not do anything.
> 
> By the way, it seems you are from Huawei. And the usage in this [official article](https://mirrors.huaweicloud.com/mirrorDetail/5ea14e7b02ea0eda6bd1e8fa) seems have a little mistake. The config command should be `sed -i "s@https://dl-cdn.alpinelinux.org/@https://repo.huaweicloud.com/@g" /etc/apk/repositories` . The comments in this article had point out this mistake, but not correct yet. :-(

Thanks, I will convey the mistake but I'm not responsible for that article actually: )

**Comment by Monokaix** [comment]
> > > Update: [this commit](https://github.com/volcano-sh/volcano/pull/3142/commits/a32b344f9990b106f6317d0449c2bcb9b13c44ba) use alpine mirror to speed up building webhook-manager docker image. I think most of Volcano's user in mainland. This mirror could be useful in improving building container image.
> > 
> > 
> > It's a good intention here. But volcano is used all over the world, we should not specify cdn url in a hard code way, my opinion is that set a parameter like env to let user specify custom cdn url or just disable it: )
> 
> You are right. I add a argument like `--build-arg APK_MIRROR="https://mirrors.huaweicloud.com"` to let users do their choice, default will not do anything.
> 
> By the way, it seems you are from Huawei. And the usage in this [official article](https://mirrors.huaweicloud.com/mirrorDetail/5ea14e7b02ea0eda6bd1e8fa) seems have a little mistake. The config command should be `sed -i "s@https://dl-cdn.alpinelinux.org/@https://repo.huaweicloud.com/@g" /etc/apk/repositories` . The comments in this article had point out this mistake, but not correct yet. :-(

You can also make suggestions in the forum https://bbs.huaweicloud.com/suggestion/new directly.

**Comment by 7sunarni** [comment]
> > > Does this cache optimization you mentioned only take effect when building multiple binaries?
> > 
> > 
> > I think when your change not including go.mod and go.sum file and just including some go file, in docker building stage, this will not download dependency. Not only in building multiple binaries, and in local develop/test stage building image.
> > > Seems that when we execute make vc-controller-manager, the go mod download will be executed, and after your pr, this step has been moved to the front, so what's the difference here?
> > 
> > 
> > As I know, docker will use the cache if the file don't change. In most case, `go.mod` and `go.sum` will not change, so docker will use cache in step3 and step4 . This is also what [Docker document](https://docs.docker.com/language/golang/build-images/#multi-stage-builds) recommend us to do. And if we just copy `.volcano` to the build, this step has file change, docker will not use cache from this.
> 
> Got that! And please rebase and make a clean pr.

Done

**Comment by Monokaix** [comment]
> --build-arg APK_MIRROR="https://mirrors.huaweicloud.com"

Is "--build-arg APK_MIRROR="https://mirrors.huaweicloud.com" updated? I didn't see that yet: )

**Comment by william-wang** [comment]
retrigger the ci.

**Comment by 7sunarni** [comment]
> > --build-arg APK_MIRROR="https://mirrors.huaweicloud.com"
> 
> Is "--build-arg APK_MIRROR="https://mirrors.huaweicloud.com" updated? I didn't see that yet: )

I don't add this parameter in Makefile. If this parameter is missing, it will not change the alpine mirror.

**Comment by Monokaix** [comment]
> > > --build-arg APK_MIRROR="https://mirrors.huaweicloud.com"
> > 
> > 
> > Is "--build-arg APK_MIRROR="https://mirrors.huaweicloud.com" updated? I didn't see that yet: )
> 
> I don't add this parameter in Makefile. If this parameter is missing, it will not change the alpine mirror.

Can the env transfer to docker image context？Seems that the env should be transferred through build-arg.

**Comment by 7sunarni** [comment]
> > > > --build-arg APK_MIRROR="https://mirrors.huaweicloud.com"
> > > 
> > > 
> > > Is "--build-arg APK_MIRROR="https://mirrors.huaweicloud.com" updated? I didn't see that yet: )
> > 
> > 
> > I don't add this parameter in Makefile. If this parameter is missing, it will not change the alpine mirror.
> 
> Can the env transfer to docker image context？Seems that the env should be transferred through build-arg.

It can. I add parameter to the Makefile.

**Review by william-wang — APPROVED** [review]
/approve

**Comment by 7sunarni** [comment]
@volcano-sh-bot  CI

**Pull request merged**